### DISCUSSION TITLE
How to deploy a flask socket io application on IIS server? step by step process

### DISCUSSION THREAD

**Discussion by kthinnal** [miguelgrinberg/Flask-SocketIO#1676] [Q&A]
Hi ,

I tried to host a basic example Flask socketIO app on the IIS server using the steps mentioned in the post 
https://medium.com/@dpralay07/deploy-a-python-flask-application-in-iis-server-and-run-on-machine-ip-address-ddb81df8edf3

deployment was successful. I  was able to browse the default endpoint, but when tried to connect from a c# socketio client application it was just kept on reconnecting, but the same code worked when I run both the applications from visual studio 

I referred global environment for my flask application so i don't understand why the client is unable to connect, i tried to log the errors but it did not log anything, 

I am new to python, could you please help me as I want to use flask sockets io in my project for implementing API

installed following 
1. eventlet latest version
2. flasksocketio  latest version

sample Flask Server app,

import eventlet

from flask import Flask
from flask_socketio import SocketIO,send,emit

app= Flask(__name__)
app.config['SECRET_KEY']="mysecret"

socketio= SocketIO(app,cors_allowed_origins="*",async_mode="eventlet")

@app.route('/')
def index():
    return "Flask Socket IO server started";

def complex_calculation():
    print('Started calculating...')    
    [x**2 for x in range(20000000)]    

@socketio.on_error()
def error_handler(e):    
    print(e)

@socketio.on("connect")
def connect():
    print( "successfully user connected to the server")

@socketio.on('myevent')
def myevent(msg):
    print('message from client : '+msg)
    complex_calculation()
    emit("myevent"," hey client, please this is your response from server",broadcast=False)

if __name__=='__main__':
    socketio.run(app,host='127.0.0.1',port=5000,log_output=True)

-------------------------------------------------------------------------------------------------------------------------
c# client that i have been using to connect 

using SocketIOClient;
using System;
using System.Diagnostics;
using System.Text;
using System.Threading.Tasks;

namespace SoketIODotnetClient
{
    class Program
    {
        static DateTime startTime = DateTime.Now;
        static async Task Main(string[] args)
        {
            Console.OutputEncoding = Encoding.UTF8;
            Trace.Listeners.Add(new TextWriterTraceListener(Console.Out));

            var uri = new Uri("http://127.0.0.1:5000/");
            var socket = new SocketIO(uri);            

            socket.OnConnected += Socket_OnConnected;
            socket.OnError += Socket_OnError;
            socket.OnPing += Socket_OnPing;
            socket.OnPong += Socket_OnPong;
            socket.OnDisconnected += Socket_OnDisconnected;
            socket.OnReconnectAttempt += Socket_OnReconnecting;
            startTime = DateTime.Now;
            
           await socket.ConnectAsync();

            await socket.EmitAsync("myevent", "hey server ....");
            socket.On("myevent", (data) =>
            {
                DateTime endTime = DateTime.Now;
                Console.WriteLine("got response message from server!!");
                Console.WriteLine(data);
                Console.WriteLine("total time taken: ", (endTime - startTime).TotalSeconds);
                socket.DisconnectAsync();
            });

            Console.ReadLine();
        }

        private static void Socket_OnError(object sender, string e)
        {
            Console.WriteLine(sender);
            Console.WriteLine(e);
        }

        private static void Socket_OnReconnecting(object sender, int e)
        {
            Console.WriteLine($"{DateTime.Now} Reconnecting: attempt = {e}");
        }

        private static void Socket_OnDisconnected(object sender, string e)
        {
            Console.WriteLine("disconnect: " + e);
        }

        private static async void Socket_OnConnected(object sender, EventArgs e)
        {
            Console.WriteLine("Socket_OnConnected");
            var socket = sender as SocketIO;
            Console.WriteLine("Socket.Id:" + socket.Id);            
        }

        private static void Socket_OnPing(object sender, EventArgs e)
        {
            Console.WriteLine("Ping");
        }

        private static void Socket_OnPong(object sender, TimeSpan e)
        {
            Console.WriteLine("Pong: " + e.TotalMilliseconds);
        }
    }
    
}

Additional info :

did sample using Flask-SocketIO with version 4.3.1, and hosted it on IIS, but it worked fine, was able to connect to the server using c# client. 

any changes in a new version that needs any additional configuration while deployment? Please help

Thanks in advance
kthinnal

**Comment by miguelgrinberg** [comment]
I don't have any knowledge of IIS and it isn't a supported web server (or a server I'm interested in adding support for). I doubt IIS is compatible with eventlet, very likely it is not. IIS is also unlikely to have WebSocket support for Python applications.

You may get better results dropping eventlet and using the `threading` async mode, but my recommendation is that you drop IIS and use one of the supported solutions listed in the documentation.

**Reply by kthinnal** [reply]
Thank you so much for your quick reply,

sorry for deviating the topic a little further
but It would be a great help if you can answer the below questions, 
As I am new to the python and flask socketIO, it would be a great help for me to get advice from an expert like you

my use case is 

1. I am trying to build an API that takes images as input and does some image processing operations and return the output JSON back to the client
2. Multiple clients can concurrently request Server and as the server does take 2 to 3 minutes time to process, 
3. Initially I thought of a normal flask Application,  where client would poll the server for a response on a timely basis
4. But as Flask-SocketIO can respond back to the client event-based, I want to use Flask-SocketIO 
5. As the other APIs in my project are hosted on IIS, I wanted to use the same IIS as the hosting server 

my questions are 
1. Can I use Flask-SocketIO for my use case, where API takes 2 to 3 minutes to respond back 
2.  If not IIS, how to deploy flask-socketIO on the windows machine, I have gone through the documentation but I did not find any deployment strategy for hosting it on windows machine
3. the best way to achieve concurrency in this case


Thanks in advance,
kthinnal

**Reply by miguelgrinberg** [reply]
1) It depends on what type of concurrency you use in your server. For a threading web server it should be fine. For an async (eventlet, gevent) web server, you can't have heavy CPU tasks, you have to move them to another process.
2) Gunicorn works fine inside the Linux compatibility layer under Windows. Eventlet and Gevent both run natively on Windows as well.
3) In your case I would try the threading concurrency mode.

**Reply by kthinnal** [reply]
Thanks for your quick response, 

I have changed async mode from "eventlet" to "threading"
socketio= SocketIO(app,cors_allowed_origins="*",async_mode="threading")

removed "import eventlet"

Please note that I am using latest version of python, flask-SocketIO

but I got the below warning message and client could not connect from then on server it has been showing 400 bad request
warning message:
WebSocket transport not available. Install simple-websocket for improved performance.
 * Serving Flask app 'app' (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
 Ref image

![image](https://user-images.githubusercontent.com/49855637/132342360-7ac85dba-f533-46a8-9d7f-6ac179218283.png)


could you please help me on this

Thanks in advance,
kthinnal

**Reply by miguelgrinberg** [reply]
You are using old versions of some packages. Upgrade `flask-socketio`, `python-socketio` and `python-engineio` to the latest versions, and also install `simple-websocket` to get websocket support.
